ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Peras/Voting/Rules.hs

```diff
@@ -0,0 +1,398 @@
+    -- at the time the certificate was issued.
+    certRoundStart = roundStart . getPerasCertRound
+
+    -- The arrival slot number of a certificate.
+    --
+    -- NOTE: this might throw a 'PastHorizonException' if the caller does not
+    -- ensure that the arrival time is within the current realizable horizon.
+    arrivalSlot cert = slotNo
+     where
+      (slotNo, _, _) =
+        pureQuery (HF.wallclockToSlot (getArrivalTime cert))
+
+    -- Does the candidate block extend the one boosted by a certificate?
+    --
+    -- This can be trivially tested by checking whether the certificate is
+    -- within the bounds of the chain prefix leading to the candidate block.
+    --
+    -- NOTE: in the case of an extremely old certificate boosting a block
+    -- beyond the immutable prefix, this could incorrectly return false even
+    -- if the voting candidate technically extends the certificate point.
+    -- However, this a boring case that we can safely ignore. Conversely,
+    -- the case of a certificate that's too new to be voted for is covered
+    -- by using the approriate prefix of our current chain.
+    candidateExtendsCert cert =
+      AF.withinFragmentBounds certBoostedBlockHeader chainAtCandidate
+     where
+      certBoostedBlockHeader = castPoint (getPerasCertBoostedBlock cert)
+
+-- | Reason for voting being disallowed
+newtype NoVoteReason = NoVoteReason (Evidence PerasVotingRule)
+  deriving (Show, Explainable)
+
+-- | Reason for voting being allowed
+newtype VoteReason = VoteReason (Evidence PerasVotingRule)
+  deriving (Show, Explainable)
+
+-- | Evaluate whether voting is allowed or not according to the voting rules
+isPerasVotingAllowed ::
+  HasPerasCertRound cert =>
+  PerasVotingView cert ->
+  Either NoVoteReason VoteReason
+isPerasVotingAllowed pvv =
+  bimap NoVoteReason VoteReason $
+    evalPred (perasVotingRules pvv)
+
+{-------------------------------------------------------------------------------
+  Voting rules
+-------------------------------------------------------------------------------}
+
+-- | Voting rules
+--
+-- Each constructor corresponds to a voting rule as per CIP-0140.
+--
+-- * VR1x correspond to the "happy" path, i.e., when nodes proceed to vote
+-- normally, whereas
+-- * VR2x correspond to the "cool-down" path, i.e., when nodes are exiting a
+-- cool-down period.
+data PerasVotingRule = VR1A | VR1B | VR2A | VR2B
+  deriving (Show, Eq)
+
+instance Explainable PerasVotingRule where
+  explain Shallow = \case
+    VR1A -> "VR-1A"
+    VR1B -> "VR-1B"
+    VR2A -> "VR-2A"
+    VR2B -> "VR-2B"
+  explain Deep = \case
+    VR1A ->
+      "voter has seen the certificate for the previous round in time"
+    VR1B ->
+      "the block being voted upon extends the most recently certified block"
+    VR2A ->
+      "the last certificate seen is sufficiently old"
+    VR2B ->
+      "the last certificate on chain is exactly one or more cooldown periods old"
```

Comment:
`exactly one or more` is unclear. I would say something like `is exactly k cooldown periods old, for some positive integer k`. Maybe `k` could be part of the `VR2B` constructor, with something like
```haskell
data PerasVotingRule = VR1A | VR1B | VR2A | VR2B Int
```